### PR TITLE
feat(stdlib): Reimplement `Number.gamma` and `Number.factorial`

### PR DESCRIPTION
**Pull request by spotandjake**
This pr re-implements `Number.gamma` and `Number.factorial`.

This pr is blocked on #2181 as the tests will fail due to the precision issue when computing `sin(pi * z) / gamma(1 - z)`

I think we should also open an issue to port the [c version here](https://git.musl-libc.org/cgit/musl/tree/src/math/tgamma.c) in the future as this approximation provides decent accuracy but we accumulate some extra error due to the error in other approximation functions such as `sin`, this should also make things a little faster.

This pr closes: #1478

### CODE REVIEW

**Comment by ospencer** [comment]
Looks like the tests are failing.

**Comment by spotandjake** [comment]
> Looks like the tests are failing.

We need to wait for #2181 to be merged.

**Comment by spotandjake** [comment]
> Looks like the tests are failing.

tests on this are now passing and its good to be merged

File: stdlib/number.gr [hunk 1]
```diff
@@ -1118,3 +1118,78 @@ provide let tan = (radians: Number) => {
     WasmI32.toGrain(newFloat64(value)): Number
   }
 }
+
+// Math.gamma implemented using the Lanczos approximation
+// https://en.wikipedia.org/wiki/Lanczos_approximation
+/**
+ * Computes the gamma function of a value using Lanczos approximation.
```
Comment (ospencer) [inline stdlib/number.gr:1125]:
```suggestion
 * Computes the gamma function of a value using the Lanczos approximation.
```

File: stdlib/number.gr [hunk 2]
```diff
@@ -1118,3 +1118,78 @@ provide let tan = (radians: Number) => {
...
+
+      let t = z + g + 0.5
+      output = sqrt(2 * pi) * (t ** (z + 0.5)) * exp(t * -1) * x
+    }
+    if (abs(output) == Infinity) Infinity else output
+  }
+}
+
+/**
+ * Computes the product of consecutive integers for an integer input and computes the gamma function for non-integer inputs.
```
Comment (ospencer) [inline stdlib/number.gr:1176]:
```suggestion
 * Computes the factorial of an integer input or the gamma function of a non-integer input.
```

File: stdlib/number.gr [hunk 3]
```diff
@@ -1118,3 +1118,78 @@ provide let tan = (radians: Number) => {
...
   }
 }
+
+// Math.gamma implemented using the Lanczos approximation
+// https://en.wikipedia.org/wiki/Lanczos_approximation
+/**
+ * Computes the gamma function of a value using Lanczos approximation.
+ *
+ * @param z: The value to interpolate
+ * @returns The gamma of the given value
```
Comment (ospencer) [inline stdlib/number.gr:1128]: Examples?

File: stdlib/number.gr [hunk 4]
```diff
@@ -1118,3 +1118,78 @@ provide let tan = (radians: Number) => {
...
+    }
+    if (abs(output) == Infinity) Infinity else output
+  }
+}
+
+/**
+ * Computes the product of consecutive integers for an integer input and computes the gamma function for non-integer inputs.
+ *
+ * @param n: The value to factorialize
+ * @returns The factorial of the given value
```
Comment (ospencer) [inline stdlib/number.gr:1179]: Examples?

**Review by ospencer — APPROVED** [review]

**Pull request merged**